Marlin/Marlin_main.cpp

```diff
@@ -1287,7 +1288,25 @@ static void update_software_endstops(AxisEnum axis) {
   {
     sw_endstop_min[axis] = base_min_pos(axis) + offs;
     sw_endstop_max[axis] = base_max_pos(axis) + offs;
+
+    #if ENABLED(AUTO_BED_LEVELING_FEATURE)
+      if (axis == Z_AXIS) {
```

Comment:
On reflection I still don't understand why the `zprobe_zoffset` was being applied to `sw_endstop_min[Z_AXIS]` in the clamp function. If the tip of the nozzle is the reference point, and the probe is below the nozzle, then there's no constraint on moving all the way down to where the nozzle is touching the bed. For a probe below the nozzle, no extra downward movement is needed. Of course, for those cases where the nozzle is the probe, so the nozzle needs to move "below" the bed, then I could see extending `sw_endstop_min[Z_AXIS]` to make room. But for a probe lower than the nozzle, it makes no sense, unless the coordinate system is (a) being considered from the probe's position instead of the nozzle's position, and (b) room is being left to move the nozzle down to the bed, within the probe-relative coordinate space.

The same question arises for `home_offset[Z_AXIS]`. This offset is applied to the current position from homing onward, and entirely changes the home position. The X and Y values from `home_offset` are applied normally to the `sw_endstop_min` and `sw_endstop_max`, and as far as I know there's no reason to treat Z any differently. For instance, if your Z home position was flashed with a `Z_MIN_POS` of `0`, then `home_offset[Z_AXIS]` was set to `4.0`, the next `G28` will set the Z position to `4.0`, and that should already be the value of `sw_endstop_min[Z_AXIS]`. However, with the current code in `clamp_to_software_endstops` the `sw_endstop_min[Z_AXIS]` will still be left at `Z=0`, allowing 4mm of downward movement, for no particular reason.